warnings in core/time.h (PR #4926) -- is_leap_year / days_in_month declared static' but never defined

### The problem

Warnings being received after PR 4926: https://github.com/esphome/esphome/pull/4926

Issue exists for ESP8266 and ESP32 with ESP-IDF framework, but not ESP32 with Arduino framework.

I tried to play around and figure out how to fix this but no luck.  Somehow core/time.cpp isn't being included properly?

### Which version of ESPHome has the issue?

2023.6.0

### What type of installation are you using?

pip

### Which version of Home Assistant has the issue?

n/a

### What platform are you using?

ESP8266

### Board

_No response_

### Component causing the issue

core/time.h

### Example YAML snippet

```yaml
esphome:
  name: asdf

esp8266:
  board: esp01_1m
```


### Anything in the logs that might be useful for us?

```txt
Compiling .pioenvs/asdf/FrameworkArduino/cont_util.cpp.o
In file included from src/esphome.h:26,
                 from src/main.cpp:3:
src/esphome/core/time.h:11:13: warning: 'bool esphome::is_leap_year(uint32_t)' declared 'static' but never defined [-Wunused-function]
   11 | static bool is_leap_year(uint32_t year);
      |             ^~~~~~~~~~~~
Compiling .pioenvs/asdf/FrameworkArduino/core_esp8266_app_entry_noextra4k.cpp.o
src/esphome/core/time.h:13:16: warning: 'uint8_t esphome::days_in_month(uint8_t, uint16_t)' declared 'static' but never defined [-Wunused-function]
   13 | static uint8_t days_in_month(uint8_t month, uint16_t year);
      |                ^~~~~~~~~~~~~
```


### Additional information

_No response_

@ssieb it happens also on ESP32 with esp-idf framework.

```
Compiling .pioenvs/esp07/src/esphome/components/wifi/wifi_component_esp_idf.o
In file included from src/esphome/components/time/real_time_clock.h:8,
                 from src/esphome/components/time/real_time_clock.cpp:1:
src/esphome/core/time.h:11:13: warning: 'bool esphome::is_leap_year(uint32_t)' declared 'static' but never defined [-Wunused-function]
 static bool is_leap_year(uint32_t year);
             ^~~~~~~~~~~~
src/esphome/core/time.h:13:16: warning: 'uint8_t esphome::days_in_month(uint8_t, uint16_t)' declared 'static' but never defined [-Wunused-function]
 static uint8_t days_in_month(uint8_t month, uint16_t year);
                ^~~~~~~~~~~~~
```

OK, good title change.  The problem is definitely a general one and not specific to total_daily_energy.  Even just using this as the entire config still gives the same warnings.

YAML config:
```
esphome:
  name: asdf

esp8266:
  board: esp01_1m
```

Warnings:
```
Compiling .pioenvs/asdf/FrameworkArduino/cont_util.cpp.o
In file included from src/esphome.h:26,
                 from src/main.cpp:3:
src/esphome/core/time.h:11:13: warning: 'bool esphome::is_leap_year(uint32_t)' declared 'static' but never defined [-Wunused-function]
   11 | static bool is_leap_year(uint32_t year);
      |             ^~~~~~~~~~~~
Compiling .pioenvs/asdf/FrameworkArduino/core_esp8266_app_entry_noextra4k.cpp.o
src/esphome/core/time.h:13:16: warning: 'uint8_t esphome::days_in_month(uint8_t, uint16_t)' declared 'static' but never defined [-Wunused-function]
   13 | static uint8_t days_in_month(uint8_t month, uint16_t year);
      |                ^~~~~~~~~~~~~
```

--- a/esphome/core/time.cpp
+++ b/esphome/core/time.cpp
@@ -2,6 +2,16 @@
 
 namespace esphome {
 
+static bool is_leap_year(uint32_t year) { return (year % 4) == 0 && ((year % 100) != 0 || (year % 400) == 0); }
+
+static uint8_t days_in_month(uint8_t month, uint16_t year) {
+  static const uint8_t DAYS_IN_MONTH[] = {0, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
+  uint8_t days = DAYS_IN_MONTH[month];
+  if (month == 2 && is_leap_year(year))
+    return 29;
+  return days;
+}
+
 size_t ESPTime::strftime(char *buffer, size_t buffer_len, const char *format) {
   struct tm c_tm = this->to_c_tm();
   return ::strftime(buffer, buffer_len, format, &c_tm);
@@ -156,16 +166,6 @@ template<typename T> bool increment_time_value(T &current, uint16_t begin, uint1
     return true;
   }
   return false;
-}
-
-static bool is_leap_year(uint32_t year) { return (year % 4) == 0 && ((year % 100) != 0 || (year % 400) == 0); }
-
-static uint8_t days_in_month(uint8_t month, uint16_t year) {
-  static const uint8_t DAYS_IN_MONTH[] = {0, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
-  uint8_t days = DAYS_IN_MONTH[month];
-  if (month == 2 && is_leap_year(year))
-    return 29;
-  return days;
 }
 
 }  // namespace esphome

--- a/esphome/core/time.h
+++ b/esphome/core/time.h
@@ -8,10 +8,6 @@ namespace esphome {
 
 template<typename T> bool increment_time_value(T &current, uint16_t begin, uint16_t end);
 
-static bool is_leap_year(uint32_t year);
-
-static uint8_t days_in_month(uint8_t month, uint16_t year);
-
 /// A more user-friendly version of struct tm from time.h
 struct ESPTime {
   /** seconds after the minute [0-60]

